mayastor/src/bdev/nexus/nexus_bdev.rs

```diff
@@ -633,29 +686,140 @@ impl Nexus {
-                subsystem.resume().await.unwrap();
+            if self.pause_waiters.is_empty() {
+                if let Some(subsystem) = NvmfSubsystem::nqn_lookup(&self.name) {
+                    self.pause_state = NexusPauseState::Unpausing;
+                    subsystem.resume().await.unwrap();
+                    // The trickiest case: a new waiter appeared during nexus
+                    // unpausing. By the agreement we keep
+                    // nexus paused for the waiters, so pause
+                    // the nexus to restore status quo.
+                    if !self.pause_waiters.is_empty() {
+                        info!(
+                            "{} concurrent nexus pausing requested during unpausing, re-pausing",
+                            self.name,
+                        );
+                        subsystem.pause().await.unwrap();
+                        self.pause_state = NexusPauseState::Paused;
+                    }
+                }
             }
         }
 
+        // Keep the Nexus paused in case there are waiters.
+        if !self.pause_waiters.is_empty() {
+            let s = self.pause_waiters.pop().unwrap();
+            s.send(0).expect("Nexus pause waiter disappeared");
+        } else {
+            self.pause_state = NexusPauseState::Unpaused;
+        }
+
         Ok(())
     }
 
-    /// suspend any incoming IO to the bdev pausing the controller allows us to
+    /// Suspend any incoming IO to the bdev pausing the controller allows us to
     /// handle internal events and which is a protocol feature.
-    pub(crate) async fn pause(&self) -> Result<(), Error> {
-        if let Some(Protocol::Nvmf) = self.shared() {
-            if let Some(subsystem) = NvmfSubsystem::nqn_lookup(&self.name) {
-                subsystem.pause().await.unwrap();
+    /// In case concurrent pause requests take place, the other callers
+    /// will wait till the nexus is resumed and will continue execution
+    /// with the nexus paused once they are awakened via resume().
+    /// Note: in order to handle cuncurrent pauses properly, this function must
+    /// be called only from the master core.
+    pub(crate) async fn pause(&mut self) -> Result<(), Error> {
+        assert_eq!(Cores::current(), Cores::first());
+
+        match self.pause_state {
+            // Pause nexus if its unpaused.
+            NexusPauseState::Unpaused => {
+                self.pause_state = NexusPauseState::Pausing;
+
+                info!("{} pausing nexus", self.name);
+                if let Some(Protocol::Nvmf) = self.shared() {
+                    if let Some(subsystem) =
+                        NvmfSubsystem::nqn_lookup(&self.name)
+                    {
+                        info!(
+                            "{} pausing subsystem {}",
+                            self.name,
+                            subsystem.get_nqn()
+                        );
+                        subsystem.pause().await.unwrap();
```

Comment:
while one future enters the await, another can also get there and not the "_" case below?

Reply:
Nope.
We get here only if the state is Unpaused and we mark it as Unpausing, so the other futures can reach this branch when we enter await.
Also, we allow executing this function only on the first core always (see asserts in the beginning), so no concurrent executions are possible.

Reply:
oh I missed the set of the pausing status above, then yes that's correct :+1: